lib/Sema/CodeSynthesis.cpp

```diff
@@ -1608,3 +1609,32 @@ ConstructorDecl *SynthesizeTypeWrappedTypeMemberwiseInitializer::evaluate(
   ctor->setBody(body, AbstractFunctionDecl::BodyKind::Parsed);
   return ctor;
 }
+
+FuncDecl *ValueDecl::getHasSymbolQueryDecl() const {
+  return evaluateOrDefault(getASTContext().evaluator,
+                           SynthesizeHasSymbolQueryRequest{this}, nullptr);
+}
+
+FuncDecl *
+SynthesizeHasSymbolQueryRequest::evaluate(Evaluator &evaluator,
+                                          const ValueDecl *decl) const {
+  auto &ctx = decl->getASTContext();
+  auto dc = decl->getModuleContext();
+
+  Mangle::ASTMangler mangler;
+  auto mangledName = ctx.AllocateCopy(mangler.mangleHasSymbolQuery(decl));
+
+  ParameterList *params = ParameterList::createEmpty(ctx);
+
+  DeclName funcName =
+      DeclName(ctx, DeclBaseName(ctx.getIdentifier(mangledName)),
+               /*argumentNames=*/ArrayRef<Identifier>());
+
+  auto i1 = BuiltinIntegerType::get(1, ctx);
+  FuncDecl *func = FuncDecl::createImplicit(
+      ctx, swift::StaticSpellingKind::None, funcName, SourceLoc(),
+      /*async=*/false, /*throws=*/false, nullptr, params, i1, dc);
+
+  func->getAttrs().add(new (ctx) SILGenNameAttr(mangledName, IsImplicit));
+  return func;
+}
```

Comment:
Would it help the inlinable case to write that synthesized function in the swiftinterface (and swiftmodule) as internal?

Reply:
It doesn't help directly. The function body cannot be spelled in Swift (or even SIL) today so we can't emit the implementation of the helper function. Emitting just the function declaration with some kind of indication that it needs special treatment is doable but it also doesn't help with the crux of the problem, which is reconstituting the `VarDecl *` that we need to pass down to IRGen.

These are my ideas for lifting the restriction in the future:
1. Invent a new SIL instruction for the `#_hasSymbol()` condition that, when serialized, serializes the `VarDecl *` so that we can re-derive the helper function.
2. Make SIL capable of representing the logic of the helper function, in which case we generate the helper function in SILGen instead of IRGen.